In [2]:
!pip install tensorflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 572.6/572.6 MB 765.0 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 78.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 80.8 MB/s eta 0:00:00
  Attempting uninstall: h5py
    Found existing installation: h5py 3.16.0
    Uninstalling h5py-3.16.0:
      Successfully uninstalled h5py-3.16.0


## **Loading library**

In [3]:
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
from tensorflow.keras.utils import to_categorical

/usr/local/lib/python3.12/dist-packages/jax/_src/cloud_tpu_init.py:86: UserWarning: Transparent hugepages are not enabled. TPU runtime startup and shutdown time should be significantly improved on TPU v5e and newer. If not already set, you may need to enable transparent hugepages in your VM image (sudo sh -c "echo always > /sys/kernel/mm/transparent_hugepage/enabled")
  warnings.warn(


## **Load the MNIST dataset:**

In [12]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [13]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


Il y a 60000 images dans les données de test et 10000 images dans les données d'entraînements avec une résolution de 28x28

## **Preprocess the data for a Fully Connected Neural Network:**

In [14]:
X_train_flat = X_train.reshape((60000, 28 * 28))
X_test_flat = X_test.reshape((10000, 28 * 28))

In [15]:
X_train_flat = X_train_flat.astype('float32') / 255.0 #Normalisation des images pour conserver l'essentielle
X_test_flat = X_test_flat.astype('float32') / 255.0

In [16]:
y_train_ho = to_categorical(y_train) #Catégorisation de la variable cible
y_test_ho = to_categorical(y_test)

## **Build and train a Fully Connected Neural Network:**

In [20]:
model = Sequential([
    Dense(128, activation='relu', input_shape=(28 * 28,)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [21]:
model.fit(X_train_flat, y_train_ho, epochs=10, batch_size=32, validation_split=0.2) #Entrainement du modèle avec 10 époques

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9212 - loss: 0.2727 - val_accuracy: 0.9516 - val_loss: 0.1584
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9646 - loss: 0.1168 - val_accuracy: 0.9651 - val_loss: 0.1140
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9756 - loss: 0.0784 - val_accuracy: 0.9673 - val_loss: 0.1115
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9819 - loss: 0.0583 - val_accuracy: 0.9718 - val_loss: 0.0917
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9857 - loss: 0.0452 - val_accuracy: 0.9721 - val_loss: 0.0979
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9877 - loss: 0.0385 - val_accuracy: 0.9734 - val_loss: 0.0998
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9900 - loss: 0.0299 - val_accuracy: 0.9760 - val_loss: 0.0917
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9916 - loss: 0.0251 - 

In [30]:
loss, accuracy = model.evaluate(X_test_flat, y_test_ho)
print(f'Test accuracy: {round(accuracy * 100,2)}%')
print(f'Test loss: {round(loss,4)}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 933us/step - accuracy: 0.9741 - loss: 0.1095
Test accuracy: 97.41%
Test loss: 0.1095


## **Preprocess the data for a Convolutional Neural Network:**

In [22]:
X_train_cnn = X_train.reshape((60000, 28, 28, 1))
X_test_cnn = X_test.reshape((10000, 28, 28, 1))

In [23]:
X_train_cnn = X_train_cnn.astype('float32') / 255.0
X_test_cnn = X_test_cnn.astype('float32') / 255.0

In [24]:
print(X_train_cnn.shape)
print(X_test_cnn.shape)

(60000, 28, 28, 1)
(10000, 28, 28, 1)


## **Build and train a Convolutional Neural Network:**

In [25]:
model_cnn = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),

    Flatten(),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

model_cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [26]:
model_cnn.fit(X_train_cnn, y_train_ho, epochs=10, batch_size=32, validation_split=0.2)

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 13s 8ms/step - accuracy: 0.9434 - loss: 0.1904 - val_accuracy: 0.9746 - val_loss: 0.0881
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9808 - loss: 0.0634 - val_accuracy: 0.9793 - val_loss: 0.0660
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9872 - loss: 0.0414 - val_accuracy: 0.9843 - val_loss: 0.0559
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9902 - loss: 0.0305 - val_accuracy: 0.9833 - val_loss: 0.0548
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9932 - loss: 0.0208 - val_accuracy: 0.9851 - val_loss: 0.0556
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9946 - loss: 0.0164 - val_accuracy: 0.9842 - val_loss: 0.0599
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9961 - loss: 0.0124 - val_accuracy: 0.9827 - val_loss: 0.0677
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 12s 8ms/step - accuracy: 0.9970 - loss: 0

In [31]:
loss, accuracy = model_cnn.evaluate(X_test_cnn, y_test_ho)
print(f'Test accuracy: {round(accuracy * 100,2)}%')
print(f'Test loss: {round(loss,2)}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9847 - loss: 0.0617
Test accuracy: 98.47%
Test loss: 0.06


## **Compare the performance:**

Dans les deux, les modèles ci ont un accuracy suppérieure aux ann (97% pour les ANN et 98% pour les CNN)

Les réseaux de neurones fully connected (ANN) traitent une image comme un simple vecteur de pixels, sans prendre en compte la structure spatiale (relations entre pixels voisins). Cela entraîne une perte d’information importante sur les formes, les contours et les textures. À l’inverse, les CNN exploitent des filtres locaux appliqués par régions (patchs), ce qui permet de préserver et d’extraire les caractéristiques spatiales de l’image. Grâce au partage de poids et à la convolution, les CNN sont donc plus efficaces pour comprendre les motifs visuels que les ANN classiques.